File: src/main/java/com/company/erp/features/chats/presence/PresenceService.java
package com.company.erp.features.chats.presence;

import com.company.erp.features.chats.ws.ChatBroadcaster;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

/**
 * In-memory presence tracker. Sessions are added on STOMP CONNECT and removed
 * on DISCONNECT. Busy-state is driven separately by {@code ChatCallService}
 * (a user is BUSY whenever they're in an active call).
 *
 * <p>State is process-local — multi-instance deployments need to move this
 * to Redis (pub/sub for cross-instance fan-out plus a hash for the map).
 * Single-instance is the current scope.</p>
 */
@Service
public class PresenceService {

    /** userId → set of currently active STOMP session ids. */
    private final Map<Long, Set<String>> sessionsByUser = new ConcurrentHashMap<>();
    /** sessionId → userId, used for O(1) cleanup on DISCONNECT. */
    private final Map<String, Long> userBySession = new ConcurrentHashMap<>();
    /** userId → last time they went OFFLINE (for "last seen 5 min ago" UI). */
    private final Map<Long, Instant> lastSeenAt = new ConcurrentHashMap<>();
    /** userId → true if they're currently in an active call. */
    private final Set<Long> busy = ConcurrentHashMap.newKeySet();
    /**
     * userId → true if the app sent an explicit "I minimized" beacon
     * ({@code POST /chats/presence/background}). Flips the user OFFLINE
     * instantly for call-routing instead of waiting ~20-30s for the STOMP
     * heartbeat to notice the OS-suspended socket. This is what lets a
     * just-minimized callee receive the VoIP/CallKit ring (the ring gate in
     * {@code ChatCallService.start} keys off OFFLINE). Cleared on the
     * foreground beacon or on the next STOMP CONNECT (reconnect ⇒ foreground).
     */
    private final Set<Long> backgrounded = ConcurrentHashMap.newKeySet();

    private final ChatBroadcaster broadcaster;
    private final SimpMessagingTemplate template;

    public PresenceService(ChatBroadcaster broadcaster, SimpMessagingTemplate template) {
        this.broadcaster = broadcaster;
        this.template = template;
    }

    public void connect(Long userId, String sessionId) {
        if (userId == null || sessionId == null) return;
        PresenceStatus before = statusOf(userId);
        // NOTE: do NOT clear the `backgrounded` override here. A STOMP CONNECT
        // is NOT proof of foreground: a KILLED iOS app woken by a call VoIP
        // push cold-starts and connects STOMP while still in the background
        // (no UI on screen). If we cleared `backgrounded` on every connect, that
        // cold-start would mark the user ONLINE, so a FOLLOW-UP call would skip
        // the apn/VoIP ring (the "2nd call: no ring after a killed-app reject"
        // bug). Only an EXPLICIT foreground beacon (`/presence/foreground` →
        // clearBackgrounded), which the client sends on a genuine
        // AppLifecycleState.resumed, means the app is really on screen — that is
        // what clears the override. A background STOMP reconnect now correctly
        // keeps the user OFFLINE so the next call still rings via apn.
        sessionsByUser.computeIfAbsent(userId, k -> ConcurrentHashMap.newKeySet()).add(sessionId);
        userBySession.put(sessionId, userId);
        PresenceStatus after = statusOf(userId);
        if (before != after) emit(userId, after);
    }

    public void disconnect(String sessionId) {
        Long userId = userBySession.remove(sessionId);
        if (userId == null) return;
        Set<String> sessions = sessionsByUser.get(userId);
        if (sessions != null) {
            sessions.remove(sessionId);
            if (sessions.isEmpty()) sessionsByUser.remove(userId);
        }
        if (!sessionsByUser.containsKey(userId)) {
            lastSeenAt.put(userId, Instant.now());
        }
        // Always emit on disconnect: status may have flipped to OFFLINE,
        // or stayed ONLINE/BUSY because other sessions are still up.
        emit(userId, statusOf(userId));
    }

    public void markBusy(Long userId) {
        if (userId == null) return;
        boolean changed = busy.add(userId);
        if (changed) emit(userId, statusOf(userId));
    }

    public void clearBusy(Long userId) {
        if (userId == null) return;
        boolean changed = busy.remove(userId);
        if (changed) emit(userId, statusOf(userId));
    }

    public PresenceStatus statusOf(Long userId) {
        // BUSY wins (in an active call). Otherwise an explicit background
        // beacon forces OFFLINE immediately — without it, a minimized iOS
        // app's suspended socket keeps the session "ONLINE" for ~20-30s, so
        // a call placed in that window is wrongly treated as foreground and
        // the VoIP/CallKit ring is skipped (the "minimized: no ring" bug).
        if (busy.contains(userId)) return PresenceStatus.BUSY;
        if (backgrounded.contains(userId)) return PresenceStatus.OFFLINE;
        return sessionsByUser.containsKey(userId)
                ? PresenceStatus.ONLINE
                : PresenceStatus.OFFLINE;
    }

    /**
     * App-lifecycle beacon: the user minimized. Flip them OFFLINE now for
     * call-routing instead of waiting for the STOMP heartbeat to time out the
     * OS-suspended socket. Idempotent.
     */
    public void markBackgrounded(Long userId) {
        if (userId == null) return;
        PresenceStatus before = statusOf(userId);
        backgrounded.add(userId);
        PresenceStatus after = statusOf(userId);
        if (before != after) {
            lastSeenAt.put(userId, Instant.now());
            emit(userId, after);
        }
    }

    /** App-lifecycle beacon: the user returned to the foreground. */
    public void clearBackgrounded(Long userId) {
        if (userId == null) return;
        PresenceStatus before = statusOf(userId);
        if (backgrounded.remove(userId)) {
            PresenceStatus after = statusOf(userId);
            if (before != after) emit(userId, after);
        }
    }

    public PresenceDto dtoOf(Long userId) {
        PresenceStatus s = statusOf(userId);
        Instant seen = (s == PresenceStatus.OFFLINE) ? lastSeenAt.get(userId) : null;
        return new PresenceDto(userId, s, seen);
    }

    /** Snapshot of every user the service has ever seen (online or offline). */
    public List<PresenceDto> snapshot() {
        Set<Long> ids = new HashSet<>();
        ids.addAll(sessionsByUser.keySet());
        ids.addAll(lastSeenAt.keySet());
        ids.addAll(busy);
        return ids.stream().map(this::dtoOf).toList();
    }

    public List<PresenceDto> dtosFor(List<Long> userIds) {
        return userIds.stream().map(this::dtoOf).toList();
    }

    private void emit(Long userId, PresenceStatus status) {
        Instant seen = (status == PresenceStatus.OFFLINE) ? lastSeenAt.get(userId) : null;
        PresenceDto dto = new PresenceDto(userId, status, seen);
        // Public topic — anyone interested in any user's presence can subscribe.
        template.convertAndSend("/topic/presence",
                Map.of("event", "presence.update", "payload", dto));
    }
}

File: app/Features/Chat/Services/PresenceService.php
<?php

namespace App\Features\Chat\Services;

use App\Features\Chat\Dto\PresenceDto;
use App\Features\Chat\Models\PresenceStatus;
use Carbon\CarbonImmutable;
use Illuminate\Support\Facades\Cache;

/**
 * Cache-backed presence tracker — port of the Spring in-memory PresenceService.
 *
 * Status resolution (identical to the Java code):
 *   - BUSY    if the user is flagged in-call (set/cleared by ChatCallService).
 *   - OFFLINE if the user sent an explicit "backgrounded" beacon.
 *   - ONLINE  if the user has at least one active session.
 *   - OFFLINE otherwise.
 *
 * The Spring service drove ONLINE/OFFLINE off STOMP CONNECT/DISCONNECT frames.
 * In this Laravel-WebSockets port the same session lifecycle is fed by
 * presence-channel membership webhooks / a client heartbeat: the websocket
 * server (or a heartbeat endpoint) calls {@link connect()} on join and
 * {@link disconnect()} on leave. Those wiring hooks are out of scope here, but
 * the service API below is complete and the BUSY logic is fully functional and
 * invoked from ChatCallService.
 *
 * State lives in the cache so it survives across worker processes. A single
 * application instance is the current scope (matching the Spring note about
 * needing Redis pub/sub for multi-instance fan-out).
 */
class PresenceService
{
    /** Cache key prefixes. */
    private const K_SESSIONS = 'chat:presence:sessions:';      // userId -> set of session ids (json array)
    private const K_SESSION_USER = 'chat:presence:sessionuser:'; // sessionId -> userId
    private const K_LAST_SEEN = 'chat:presence:lastseen:';     // userId -> epoch seconds
    private const K_BUSY = 'chat:presence:busy:';              // userId -> 1
    private const K_BACKGROUNDED = 'chat:presence:bg:';        // userId -> 1
    private const K_KNOWN = 'chat:presence:known';             // set of every userId ever seen (json array)

    /** Long TTL so presence keys effectively persist for the process lifetime. */
    private const TTL = 86400;

    /** @var ChatBroadcaster */
    private $broadcaster;

    public function __construct(ChatBroadcaster $broadcaster)
    {
        $this->broadcaster = $broadcaster;
    }

    /**
     * Register a session for a user (analogue of STOMP CONNECT).
     *
     * NOTE: like the Java code, this deliberately does NOT clear the
     * `backgrounded` override — a reconnect is not proof of foreground (a
     * killed iOS app woken by a VoIP push reconnects while still backgrounded).
     * Only an explicit foreground beacon clears it.
     */
    public function connect(int $userId, string $sessionId): void
    {
        $before = $this->statusOf($userId);

        $sessions = $this->sessions($userId);
        $sessions[$sessionId] = true;
        $this->putSessions($userId, $sessions);
        Cache::put(self::K_SESSION_USER.$sessionId, $userId, self::TTL);
        $this->remember($userId);

        $after = $this->statusOf($userId);
        if ($before !== $after) {
            $this->emit($userId, $after);
        }
    }

    /** Remove a session (analogue of STOMP DISCONNECT). */
    public function disconnect(string $sessionId): void
    {
        $userId = Cache::get(self::K_SESSION_USER.$sessionId);
        if ($userId === null) {
            return;
        }
        $userId = (int) $userId;
        Cache::forget(self::K_SESSION_USER.$sessionId);

        $sessions = $this->sessions($userId);
        unset($sessions[$sessionId]);
        if (empty($sessions)) {
            Cache::forget(self::K_SESSIONS.$userId);
            Cache::put(self::K_LAST_SEEN.$userId, CarbonImmutable::now('UTC')->getTimestamp(), self::TTL);
        } else {
            $this->putSessions($userId, $sessions);
        }

        // Always emit on disconnect: status may have flipped to OFFLINE, or
        // stayed ONLINE/BUSY because other sessions are still up.
        $this->emit($userId, $this->statusOf($userId));
    }

    /** Flag the user as in an active call (BUSY wins over ONLINE/OFFLINE). */
    public function markBusy(int $userId): void
    {
        if (Cache::has(self::K_BUSY.$userId)) {
            return;
        }
        Cache::put(self::K_BUSY.$userId, 1, self::TTL);
        $this->remember($userId);
        $this->emit($userId, $this->statusOf($userId));
    }

    /** Clear the user's in-call flag. */
    public function clearBusy(int $userId): void
    {
        if (! Cache::has(self::K_BUSY.$userId)) {
            return;
        }
        Cache::forget(self::K_BUSY.$userId);
        $this->emit($userId, $this->statusOf($userId));
    }

    public function statusOf(int $userId): string
    {
        if (Cache::has(self::K_BUSY.$userId)) {
            return PresenceStatus::BUSY;
        }
        if (Cache::has(self::K_BACKGROUNDED.$userId)) {
            return PresenceStatus::OFFLINE;
        }
        return ! empty($this->sessions($userId))
            ? PresenceStatus::ONLINE
            : PresenceStatus::OFFLINE;
    }

    /**
     * App-lifecycle beacon: the user minimized. Flip them OFFLINE now for
     * call-routing instead of waiting for a heartbeat to time out the suspended
     * socket. Idempotent.
     */
    public function markBackgrounded(int $userId): void
    {
        $before = $this->statusOf($userId);
        Cache::put(self::K_BACKGROUNDED.$userId, 1, self::TTL);
        $this->remember($userId);
        $after = $this->statusOf($userId);
        if ($before !== $after) {
            Cache::put(self::K_LAST_SEEN.$userId, CarbonImmutable::now('UTC')->getTimestamp(), self::TTL);
            $this->emit($userId, $after);
        }
    }

    /** App-lifecycle beacon: the user returned to the foreground. */
    public function clearBackgrounded(int $userId): void
    {
        $before = $this->statusOf($userId);
        if (Cache::has(self::K_BACKGROUNDED.$userId)) {
            Cache::forget(self::K_BACKGROUNDED.$userId);
            $after = $this->statusOf($userId);
            if ($before !== $after) {
                $this->emit($userId, $after);
            }
        }
    }

    /**
     * @return array<string, mixed>
     */
    public function dtoOf(int $userId): array
    {
        $status = $this->statusOf($userId);
        $seen = $status === PresenceStatus::OFFLINE ? $this->lastSeen($userId) : null;
        return PresenceDto::from($userId, $status, $seen);
    }

    /**
     * Snapshot of every user the service has ever seen (online or offline).
     *
     * @return array<int, array<string, mixed>>
     */
    public function snapshot(): array
    {
        $out = [];
        foreach ($this->known() as $userId) {
            $out[] = $this->dtoOf((int) $userId);
        }
        return $out;
    }

    /**
     * @param array<int, int> $userIds
     * @return array<int, array<string, mixed>>
     */
    public function dtosFor(array $userIds): array
    {
        $out = [];
        foreach ($userIds as $userId) {
            $out[] = $this->dtoOf((int) $userId);
        }
        return $out;
    }

    // --- internals ----------------------------------------------------------

    private function emit(int $userId, string $status): void
    {
        $seen = $status === PresenceStatus::OFFLINE ? $this->lastSeen($userId) : null;
        $this->broadcaster->presence(PresenceDto::from($userId, $status, $seen));
    }

    private function lastSeen(int $userId): ?CarbonImmutable
    {
        $ts = Cache::get(self::K_LAST_SEEN.$userId);
        return $ts === null ? null : CarbonImmutable::createFromTimestampUTC((int) $ts);
    }

    /**
     * @return array<string, bool>
     */
    private function sessions(int $userId): array
    {
        $raw = Cache::get(self::K_SESSIONS.$userId);
        return is_array($raw) ? $raw : [];
    }

    /**
     * @param array<string, bool> $sessions
     */
    private function putSessions(int $userId, array $sessions): void
    {
        Cache::put(self::K_SESSIONS.$userId, $sessions, self::TTL);
    }

    private function remember(int $userId): void
    {
        $known = $this->known();
        if (! in_array($userId, $known, true)) {
            $known[] = $userId;
            Cache::put(self::K_KNOWN, $known, self::TTL);
        }
    }

    /**
     * @return array<int, int>
     */
    private function known(): array
    {
        $raw = Cache::get(self::K_KNOWN);
        return is_array($raw) ? array_map('intval', $raw) : [];
    }
}
